### PR TITLE
Lock SCSS so we only run 1 job at a time

### PR DESCRIPTION
**Pull request by rullzer**
This is bit hacky but a start to lock the SCSS compiler properly

Right now it will just fail if the file is locked. This is suboptimal. But for the POC this is fine IMO.

Todo:
* [x] Busy wait for 10 seconds max until it is done
* [ ] Try to cleanup initialization

@ChristophWurst @MorrisJobke as discussed

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Has anybody been able to reproduce this locally yet?

**Comment by skjnldsv** [comment]
@ChristophWurst I tried running multiple same instance after a css cache clear and I got no issue whatsoever :)
I could check with a debug log if this is really triggered properly though.

**Comment by ChristophWurst** [comment]
> I could check with a debug log if this is really triggered properly though.

That would be awesome!

**Comment by skjnldsv** [comment]
I fixed a few stuff.
In the end we should also probably lock the css cache deletion.
If an admin change a variable value and two people refreshes while the cache is being deleted, the cache will be deleted once again.

Also, why is deleting the cache taking so long? @rullzer @nickvergessen @juliushaertl  ?

File: lib/private/Template/SCSSCacher.php [hunk 1]
```diff
@@ -289,7 +327,7 @@ private function cache(string $path, string $fileNameCSS, string $fileNameSCSS,
 				'@import "functions.scss";' .
 				'@import "' . $fileNameSCSS . '";');
 		} catch (ParserException $e) {
-			$this->logger->error($e, ['app' => 'core']);
+			$this->logger->error($e->getMessage(), ['app' => 'core']);
```
Comment (kesselb) [inline lib/private/Template/SCSSCacher.php]:
```suggestion
			$this->logger->logException($e, ['app' => 'core']);
```

**Review by ChristophWurst — APPROVED** [review]

File: lib/private/Template/SCSSCacher.php [hunk 2]
```diff
@@ -358,7 +401,7 @@ private function getInjectedVariables(): string {
 			$scss->compile($variables);
 			$this->injectedVariables = $variables;
 		} catch (ParserException $e) {
-			$this->logger->error($e, ['app' => 'core']);
+			$this->logger->error($e->getMessage(), ['app' => 'core']);
```
Comment (kesselb) [inline lib/private/Template/SCSSCacher.php]:
```suggestion
			$this->logger->logException($e, ['app' => 'core']);
```

I missed that one last time :see_no_evil:

Comment (skjnldsv) [reply lib/private/Template/SCSSCacher.php]: :heart:

**Comment by skjnldsv** [comment]
@rullzer for you :)

**Review by georgehrke — APPROVED** [review]

**Comment by MorrisJobke** [comment]
```
130 There was 1 failure:
131	
132	1) Test\Template\SCSSCacherTest::testResetCache
133	OCP\ICacheFactory::createDistributed('SCSS-cached-'): OCP\ICache was not expected to be called more than once.
134	
135	/drone/src/lib/private/Template/SCSSCacher.php:371
136	/drone/src/tests/lib/Template/SCSSCacherTest.php:541
```

File: lib/private/Template/SCSSCacher.php [hunk 3]
```diff
@@ -150,7 +157,35 @@ public function process(string $root, string $file, string $app): bool {
...
+			sleep(1);
+			while ($retry < 10) {
+				if (!$this->variablesChanged() && $this->isCached($fileNameCSS, $app)) {
+					// Inject icons vars css if any
+					$this->lockingCache->remove($lockKey);
+					$this->logger->debug('SCSSCacher: ' .$lockKey.' is now available after '.$retry.'s. Moving on...', ['app' => 'core']);
+					return $this->injectCssVariablesIfAny();
+				}
+				$this->logger->debug('SCSSCacher: scss cache file locked for '.$lockKey, ['app' => 'core']);
+				sleep($retry);
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]:
```suggestion
				sleep(1);
```

File: lib/private/Template/SCSSCacher.php [hunk 4]
```diff
@@ -180,19 +215,24 @@ public function getCachedCSS(string $appName, string $fileName): ISimpleFile {
 	 */
 	private function isCached(string $fileNameCSS, string $app) {
 		$key = $this->config->getSystemValue('version') . '/' . $app . '/' . $fileNameCSS;
-		if (!$this->config->getSystemValue('debug') && $cacheValue = $this->isCachedCache->get($key)) {
+
+		// If the file mtime is older than our cached one,
```
Comment (rullzer) [inline lib/private/Template/SCSSCacher.php]:
```suggestion
		// If the file mtime is more recent than our cached one,
```

File: lib/private/Template/SCSSCacher.php (hunk 3 above)
Comment (skjnldsv) [reply lib/private/Template/SCSSCacher.php]: No it was don on purpose.
SOmetimes it's not enough. Increasing the sleep was a nice way of making the user wait! :)

Comment (rullzer) [reply lib/private/Template/SCSSCacher.php]: so 55 seconds of wait?

Comment (skjnldsv) [reply lib/private/Template/SCSSCacher.php]: 45! :D

Comment (rullzer) [reply lib/private/Template/SCSSCacher.php]: But fair enough we can tweak later

Comment (skjnldsv) [reply lib/private/Template/SCSSCacher.php]: We could fail before, but then we should increase the number of retries.
I felt not trying too much was easier on the process as variablesChanged and isCached can be quite expensive (especially on object storage)

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Comment by rullzer** [comment]
Ok lets do this!

**Comment by leo-lb** [comment]
@rullzer Hello, this is causing extreme slowdowns in production because the lock never gets acquired.

Screenshot Capture - 2020-05-20 - 18-09-19

I think it's never a good idea to sleep in a PHP script because users will be waiting, this totals to 60 seconds of wait time on every single page in Nextcloud, my users are going crazy.

**Pull request merged**